src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -215,109 +115,148 @@ internal void OnSolutionBuild(object sender, UIContextChangedEventArgs e)
-                IDictionary<DocumentId, IList<DiagnosticData>> liveDocumentErrors;
-                GetLiveProjectAndDocumentErrors(out liveProjectErrors, out liveDocumentErrors);
+                // nothing to do
+                if (_state == null)
+                {
+                    return;
+                }
 
-                using (var documentIds = SharedPools.Default<List<DocumentId>>().GetPooledObject())
-                using (var projectIds = SharedPools.Default<List<ProjectId>>().GetPooledObject())
+                // we are about to update live analyzer data using one from build.
+                // pause live analyzer
+                var service = _workspace.Services.GetService<IGlobalOperationNotificationService>();
+                using (var operation = service.Start("BuildDone"))
                 {
-                    lock (_gate)
-                    {
-                        documentIds.Object.AddRange(_documentToDiagnosticsMap.Keys);
-                        projectIds.Object.AddRange(_projectToDiagnosticsMap.Keys);
-                    }
+                    // we will have a race here since we can't track version of solution the out of proc build actually used.
+                    // result of the race will be us dropping some diagnostics from the build to the floor.
+                    var solution = _workspace.CurrentSolution;
 
-                    foreach (var documentId in documentIds.Object)
-                    {
-                        var errors = liveDocumentErrors.GetValueOrDefault(documentId);
-                        RemoveBuildErrorsDuplicatedByLiveErrorsAndReport(documentId.ProjectId, documentId, errors, reportOnlyIfChanged: false);
-                    }
+                    await CleanupAllLiveErrorsIfNeededAsync(solution).ConfigureAwait(false);
 
-                    foreach (var projectId in projectIds.Object)
+                    var supportedIdMap = GetSupportedLiveDiagnosticId(solution, _state);
+                    Func<DiagnosticData, bool> liveDiagnosticChecker = d =>
                     {
-                        var errors = liveProjectErrors.GetValueOrDefault(projectId);
-                        RemoveBuildErrorsDuplicatedByLiveErrorsAndReport(projectId, null, errors, reportOnlyIfChanged: false);
-                    }
+                        // REVIEW: we probably need a better design on de-duplicating live and build errors. or don't de-dup at all.
+                        //         for now, we are special casing compiler error case.
+                        var project = solution.GetProject(d.ProjectId);
+                        if (project == null)
+                        {
+                            // project doesn't exist
+                            return false;
+                        }
+
+                        if (_diagnosticService.IsCompilerDiagnostic(project.Language, d) && d.DocumentId == null)
```

Comment:
Maybe add a comment here explaining that we think that diagnostics without locations are the only ones that are not produced by the compiler's live analysis.